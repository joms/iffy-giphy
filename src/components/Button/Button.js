import React from 'react';
import PropTypes from 'prop-types';

const Button = ({ onClick, children, disabled, ...rest }) => {
    const click = e => {
        if (!disabled) {
            onClick(e);
        }
    };

    return (
        <button onClick={click} disabled={disabled} {...rest}>
            {children}
        </button>
    );
};

Button.defaultProps = {
    onClick: () => null,
    disabled: false,
};

Button.propTypes = {
    children: PropTypes.node.isRequired,
    onClick: PropTypes.func,
    rest: PropTypes.object,
    disabled: PropTypes.bool,
};

export default Button;
