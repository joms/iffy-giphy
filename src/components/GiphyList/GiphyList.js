import React from 'react';
import PropTypes from 'prop-types';

export const ROW = 'ROW';
export const CARD = 'CARD';

const GiphyList = ({ giphs, as: As, mode }) => {
    return (
        <div className={`giphylist giphylist-${mode.toLowerCase()}`}>
            {giphs.map(gif => (
                <As key={gif.id} gif={gif} />
            ))}
        </div>
    );
};

GiphyList.defaultProps = {
    mode: CARD,
};

GiphyList.propTypes = {
    mode: PropTypes.oneOf([ROW, CARD]),
};

export default GiphyList;
