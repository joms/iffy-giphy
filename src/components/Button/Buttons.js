import React from 'react';
import PropTypes from 'prop-types';
import Button from '.';

const Buttons = ({ children }) => <div className="buttons">{children}</div>;

Buttons.propTypes = {
    children: PropTypes.node.isRequired,
};

export default Buttons;
