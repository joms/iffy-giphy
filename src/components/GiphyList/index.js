import GiphyList from './GiphyList';

export default GiphyList;
