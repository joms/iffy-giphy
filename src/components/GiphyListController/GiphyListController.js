import React, { useState, useEffect } from 'react';
import cx from 'classnames';
import Button, { Buttons } from '../Button';

import { CARD } from '../GiphyList';

const BUTTONS = {
    ROW: 'list',
    CARD: 'th',
};

const GiphyListController = ({ onChange }) => {
    const [mode, setMode] = useState(CARD);

    useEffect(() => {
        onChange(mode);
    }, [mode, onChange]);

    const onClick = _mode => e => setMode(_mode);

    return (
        <div className="giphylist-controller">
            <Buttons>
                {Object.entries(BUTTONS).map(([type, icon]) => (
                    <Button
                        key={type}
                        onClick={onClick(type)}
                        className={cx({ active: mode === type })}
                        title={`${type.toLowerCase()}-view`}
                    >
                        <i className={`fas fa-${icon}`} />
                    </Button>
                ))}
            </Buttons>
        </div>
    );
};

export default GiphyListController;
