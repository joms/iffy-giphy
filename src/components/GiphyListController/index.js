import GiphyListController from './GiphyListController';

export default GiphyListController;
