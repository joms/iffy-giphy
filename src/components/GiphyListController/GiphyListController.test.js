import React from 'react';
import { mount } from 'enzyme';
import GiphyListController from '.';
import Button from '../Button';
import { ROW, CARD } from '../GiphyList';

describe('<GiphyListController>', () => {
    const onChange = jest.fn();
    const wrapper = mount(<GiphyListController onChange={onChange} />);
    it('matches snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('switches mode', () => {
        wrapper
            .find(Button)
            .at(0)
            .simulate('click');
        expect(onChange).toHaveBeenCalledWith(ROW);
        wrapper
            .find(Button)
            .at(1)
            .simulate('click');
        expect(onChange).toHaveBeenCalledWith(CARD);
    });
});
