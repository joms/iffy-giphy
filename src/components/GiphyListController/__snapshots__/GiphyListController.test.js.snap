// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GiphyListController> matches snapshot 1`] = `
<GiphyListController
  onChange={
    [MockFunction] {
      "calls": Array [
        Array [
          "CARD",
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
>
  <div
    className="giphylist-controller"
  >
    <Buttons>
      <div
        className="buttons"
      >
        <Button
          className=""
          disabled={false}
          key="ROW"
          onClick={[Function]}
          title="row-view"
        >
          <button
            className=""
            disabled={false}
            onClick={[Function]}
            title="row-view"
          >
            <i
              className="fas fa-list"
            />
          </button>
        </Button>
        <Button
          className="active"
          disabled={false}
          key="CARD"
          onClick={[Function]}
          title="card-view"
        >
          <button
            className="active"
            disabled={false}
            onClick={[Function]}
            title="card-view"
          >
            <i
              className="fas fa-th"
            />
          </button>
        </Button>
      </div>
    </Buttons>
  </div>
</GiphyListController>
`;
